File: src/covers.rs
use std::cell::RefCell;
use std::path::PathBuf;
use std::rc::Rc;
use std::sync::mpsc::{self, Receiver, Sender};

use slint::Model;

use crate::entries::GameEntry;
use crate::GameViewData;

pub const COVER_WINDOW: usize = 2;
pub const COVER_MAX_DIM: u32 = 640;

pub fn hsl_to_rgb(hue: f32, saturation: f32, lightness: f32) -> slint::Color {
    let chroma = (1.0 - (2.0 * lightness - 1.0).abs()) * saturation;
    let hp = hue / 60.0;
    let x = chroma * (1.0 - ((hp % 2.0) - 1.0).abs());

    let sector = (hp as u32) % 6;
    let (r1, g1, b1) = match sector {
        0 => (chroma, x, 0.0),
        1 => (x, chroma, 0.0),
        2 => (0.0, chroma, x),
        3 => (0.0, x, chroma),
        4 => (x, 0.0, chroma),
        _ => (chroma, 0.0, x),
    };

    let m = lightness - chroma / 2.0;
    let r = ((r1 + m) * 255.0).round() as u8;
    let g = ((g1 + m) * 255.0).round() as u8;
    let b = ((b1 + m) * 255.0).round() as u8;

    slint::Color::from_rgb_u8(r, g, b)
}

pub fn placeholder_colors(name: &str) -> (slint::Color, slint::Color) {
    const SATURATION: f32 = 0.55;
    const TOP_LIGHTNESS: f32 = 0.30;
    const BOTTOM_LIGHTNESS: f32 = 0.08;

    let mut hash: u32 = 0;
    for b in name.bytes() {
        hash = hash.wrapping_mul(31).wrapping_add(b as u32);
    }
    let hue = (hash % 360) as f32;

    (
        hsl_to_rgb(hue, SATURATION, TOP_LIGHTNESS),
        hsl_to_rgb(hue, SATURATION, BOTTOM_LIGHTNESS),
    )
}

pub fn circular_distance(a: usize, b: usize, n: usize) -> usize {
    if n == 0 {
        return usize::MAX;
    }
    let forward = (a + n - b) % n;
    forward.min(n - forward)
}

pub fn make_view_data(game: &GameEntry, cover_art: slint::Image) -> GameViewData {
    let (c1, c2) = placeholder_colors(&game.name);
    let initial = game
        .name
        .chars()
        .next()
        .map(|c| c.to_uppercase().to_string())
        .unwrap_or_default();

    let has_cover = cover_art.size().width > 0 && cover_art.size().height > 0;

    GameViewData {
        game_id: game.id as i32,
        title: game.name.as_str().into(),
        initial: initial.as_str().into(),
        cover_art,
        has_cover,
        c1,
        c2,
    }
}

pub struct DecodedCover {
    pub width: u32,
    pub height: u32,
    pub rgba: Vec<u8>,
}

pub struct CoverState {
    games: Vec<GameEntry>,
    model: Rc<slint::VecModel<GameViewData>>,
    cache: Vec<Option<slint::Image>>,
    loading: Vec<bool>,
    current: usize,
    result_sender: Sender<(usize, Option<DecodedCover>)>,
    result_receiver: Receiver<(usize, Option<DecodedCover>)>,
    _drain_timer: slint::Timer,
}

impl CoverState {
    pub fn build_model(games: &[GameEntry]) -> Rc<slint::VecModel<GameViewData>> {
        let items: Vec<GameViewData> = games
            .iter()
            .map(|g| make_view_data(g, slint::Image::default()))
            .collect();
        Rc::new(slint::VecModel::from(items))
    }

    pub fn init(
        games: Vec<GameEntry>,
        model: Rc<slint::VecModel<GameViewData>>,
    ) -> Rc<RefCell<Self>> {
        let count = games.len();
        let (tx, rx) = mpsc::channel();
        let drain_timer = slint::Timer::default();

        let state = Rc::new(RefCell::new(Self {
            games,
            model,
            cache: vec![None; count],
            loading: vec![false; count],
            current: 0,
            result_sender: tx,
            result_receiver: rx,
            _drain_timer: drain_timer,
        }));

        let weak = Rc::downgrade(&state);
        state.borrow()._drain_timer.start(
            slint::TimerMode::Repeated,
            std::time::Duration::from_millis(16),
            move || {
                if let Some(s) = weak.upgrade() {
                    s.borrow_mut().drain_results();
                }
            },
        );

        state.borrow_mut().refresh();
        state
    }

    pub fn set_current(&mut self, index: usize) {
        self.current = index;
        self.refresh();
    }

    pub fn refresh(&mut self) {
        let n = self.games.len();
        for i in 0..n {
            let in_window = (n <= 2 * COVER_WINDOW + 1)
                || (circular_distance(i, self.current, n) <= COVER_WINDOW);

            if in_window && self.cache[i].is_none() && !self.loading[i] {
                if let Some(ref path_str) = self.games[i].cover {
                    self.loading[i] = true;
                    let path = PathBuf::from(path_str);
                    let tx = self.result_sender.clone();
                    std::thread::spawn(move || {
                        let decoded = (|| -> Option<DecodedCover> {
                            let dynamic_img = image::ImageReader::open(&path).ok()?.decode().ok()?;
                            let (w, h) = (dynamic_img.width(), dynamic_img.height());
                            let (w, h, rgba) = if w > COVER_MAX_DIM || h > COVER_MAX_DIM {
                                let resized = dynamic_img.thumbnail(COVER_MAX_DIM, COVER_MAX_DIM);
                                (resized.width(), resized.height(), resized.to_rgba8().into_raw())
                            } else {
                                (w, h, dynamic_img.to_rgba8().into_raw())
                            };
                            Some(DecodedCover { width: w, height: h, rgba })
                        })();
                        let _ = tx.send((i, decoded));
                    });
                } else {
                    self.cache[i] = Some(slint::Image::default());
                }
            } else if !in_window && self.cache[i].is_some() {
                self.cache[i] = None;
                self.loading[i] = false;
                self.model
                    .set_row_data(i, make_view_data(&self.games[i], slint::Image::default()));
            }
        }
    }

    fn drain_results(&mut self) {
        while let Ok((index, decoded)) = self.result_receiver.try_recv() {
            self.apply_cover(index, decoded);
        }
    }

    fn apply_cover(&mut self, index: usize, decoded: Option<DecodedCover>) {
        let n = self.games.len();
        if index >= self.loading.len() {
            return;
        }

        self.loading[index] = false;

        let in_window = (n <= 2 * COVER_WINDOW + 1)
            || (circular_distance(index, self.current, n) <= COVER_WINDOW);
        if !in_window {
            return;
        }
        if self.cache[index].is_some() {
            return;
        }

        let img = if let Some(d) = decoded {
            let mut pixel_buffer =
                slint::SharedPixelBuffer::<slint::Rgba8Pixel>::new(d.width, d.height);
            pixel_buffer.make_mut_bytes().copy_from_slice(&d.rgba);
            slint::Image::from_rgba8(pixel_buffer)
        } else {
            slint::Image::default()
        };

        self.cache[index] = Some(img.clone());
        self.model
            .set_row_data(index, make_view_data(&self.games[index], img));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circular_distance() {
        assert_eq!(circular_distance(0, 0, 5), 0);
        assert_eq!(circular_distance(0, 1, 5), 1);
        assert_eq!(circular_distance(0, 4, 5), 1);
        assert_eq!(circular_distance(0, 2, 5), 2);
        assert_eq!(circular_distance(0, 3, 5), 2);
    }

    #[test]
    fn test_placeholder_colors() {
        let (c1, c2) = placeholder_colors("Sonic");
        assert_ne!(c1, c2);
    }
}

File: src/covers.cpp
#include "covers.h"

#include <algorithm>
#include <cmath>
#include <numbers>
#include <thread>

#include "nothrow.h"

// --- Color utilities ---

slint::Color hsl_to_rgb(float hue, float saturation, float lightness) {
    float chroma = (1.0f - std::abs(2.0f * lightness - 1.0f)) * saturation;
    float hp = hue / 60.0f;
    float x = chroma * (1.0f - std::abs(std::fmod(hp, 2.0f) - 1.0f));

    float r1, g1, b1;
    uint32_t sector = static_cast<uint32_t>(hp) % 6;
    switch (sector) {
        case 0: r1 = chroma; g1 = x; b1 = 0; break;
        case 1: r1 = x; g1 = chroma; b1 = 0; break;
        case 2: r1 = 0; g1 = chroma; b1 = x; break;
        case 3: r1 = 0; g1 = x; b1 = chroma; break;
        case 4: r1 = x; g1 = 0; b1 = chroma; break;
        default: r1 = chroma; g1 = 0; b1 = x; break;
    }

    float m = lightness - chroma / 2.0f;
    uint8_t r = static_cast<uint8_t>(std::round((r1 + m) * 255.0f));
    uint8_t g = static_cast<uint8_t>(std::round((g1 + m) * 255.0f));
    uint8_t b = static_cast<uint8_t>(std::round((b1 + m) * 255.0f));

    return slint::Color::from_rgb_uint8(r, g, b);
}

std::pair<slint::Color, slint::Color> placeholder_colors(const std::string& name) {
    constexpr float SATURATION = 0.55f;
    constexpr float TOP_LIGHTNESS = 0.30f;
    constexpr float BOTTOM_LIGHTNESS = 0.08f;

    uint32_t hash = 0;
    for (uint8_t c : name) {
        hash = hash * 31 + c;
    }
    uint32_t hue = hash % 360;

    return {
        hsl_to_rgb(static_cast<float>(hue), SATURATION, TOP_LIGHTNESS),
        hsl_to_rgb(static_cast<float>(hue), SATURATION, BOTTOM_LIGHTNESS),
    };
}

size_t circular_distance(size_t a, size_t b, size_t n) {
    if (n == 0) return SIZE_MAX;
    size_t forward = (a + n - b) % n;
    return std::min(forward, n - forward);
}

// --- Helper to create GameViewData ---

static GameViewData make_view_data(const GameEntry& game, const slint::Image& cover_art) {
    auto [c1, c2] = placeholder_colors(game.name);

    std::string initial;
    if (!game.name.empty()) {
        char c = static_cast<char>(std::toupper(static_cast<unsigned char>(game.name[0])));
        initial = std::string(1, c);
    }

    bool has_cover = cover_art.size().width > 0 && cover_art.size().height > 0;

    GameViewData vd;
    vd.game_id = static_cast<int32_t>(game.id);
    vd.title = slint::SharedString(game.name);
    vd.initial = slint::SharedString(initial);
    vd.cover_art = cover_art;
    vd.has_cover = has_cover;
    vd.c1 = c1;
    vd.c2 = c2;
    return vd;
}

// --- CoverState implementation ---

CoverState::CoverState(const std::vector<GameEntry>& games,
                       std::shared_ptr<slint::VectorModel<GameViewData>> model)
    : games_(games)
    , model_(model)
    , cache_(games.size())
    , loading_(games.size(), false)
    , current_(0) {
}

CoverState::~CoverState() {
    {
        std::lock_guard<std::mutex> lock(threads_mutex_);
        for (auto& t : loading_threads_) {
            t.request_stop();
        }
    }
    loading_threads_.clear();
}

std::shared_ptr<slint::VectorModel<GameViewData>> CoverState::build_model(
    const std::vector<GameEntry>& games) {
    auto model = std::shared_ptr<slint::VectorModel<GameViewData>>(new(std::nothrow) slint::VectorModel<GameViewData>());
    for (const auto& game : games) {
        model->push_back(make_view_data(game, slint::Image()));
    }
    return model;
}

std::shared_ptr<CoverState> CoverState::init(
    const std::vector<GameEntry>& games,
    std::shared_ptr<slint::VectorModel<GameViewData>> model) {
    auto state = std::shared_ptr<CoverState>(new(std::nothrow) CoverState(games, model));

    state->drain_timer_.start(slint::TimerMode::Repeated,
        std::chrono::milliseconds(16),
        [weak = std::weak_ptr<CoverState>(state)]() {
            if (auto s = weak.lock()) {
                s->drain_results();
            }
        });

    state->refresh();
    return state;
}

void CoverState::set_current(size_t index) {
    current_ = index;
    refresh();
}

void CoverState::refresh() {
    size_t n = games_.size();

    for (size_t i = 0; i < n; ++i) {
        bool in_window = (n <= 2 * COVER_WINDOW + 1) ||
                         (circular_distance(i, current_, n) <= COVER_WINDOW);

        if (in_window && !cache_[i] && !loading_[i]) {
            if (games_[i].cover) {
                loading_[i] = true;
                size_t idx = i;
                std::string path = *games_[i].cover;
                auto self = shared_from_this();

                std::lock_guard<std::mutex> lock(threads_mutex_);
                loading_threads_.emplace_back([self, idx, path](std::stop_token) {
                    std::optional<slint::Image> result;

                    auto image = slint::Image::load_from_path(slint::SharedString(path));
                    if (image.size().width > 0 && image.size().height > 0) {
                        result = image;
                    }

                    std::lock_guard<std::mutex> lock(self->result_mutex_);
                    self->pending_results_.push_back({idx, result});
                });
            } else {
                cache_[i] = slint::Image();
            }
        } else if (!in_window && cache_[i]) {
            cache_[i] = std::nullopt;
            loading_[i] = false;
            model_->set_row_data(i, make_view_data(games_[i], slint::Image()));
        }
    }
}

void CoverState::drain_results() {
    std::vector<DecodeResult> results;
    {
        std::lock_guard<std::mutex> lock(result_mutex_);
        results = std::move(pending_results_);
    }

    for (auto& [index, image] : results) {
        apply_cover(index, image);
    }
}

void CoverState::apply_cover(size_t index, const std::optional<slint::Image>& image) {
    size_t n = games_.size();
    if (index >= loading_.size()) return;

    loading_[index] = false;

    bool in_window = (n <= 2 * COVER_WINDOW + 1) ||
                     (circular_distance(index, current_, n) <= COVER_WINDOW);
    if (!in_window) return;
    if (cache_[index]) return;

    cache_[index] = image.value_or(slint::Image());
    model_->set_row_data(index, make_view_data(games_[index], *cache_[index]));
}
